Possible fix for TheCBProject/EnderStorage#11

bschoenmaeckers opened the pull request:
This is a possible fix for TheCBProject/EnderStorage#11.  I don't know if the null check in inWorld() was for a reason on the player connection, I browsed all CC mods but could not find a problem.

covers1624 commented:
looks good, will only be in 1,9+ tho, but i may cherry pick stuff to go back to 1.7 if people REALLY need it.

bschoenmaeckers committed:
Fixed #11

covers1624 commented:
The thing is, why in gods name are you setting the player for a loading screen entity render..

bschoenmaeckers commented:
The answer is simple, because I have to. When you render a mob it will check if it is visible to the player, and if there in't a player.....NPE.

covers1624 commented:
doRenderEntity will not throw an NPE in 1.10 if the player instance is null, The only thing it references the player for is the debug bounding box and even then there are null checks.

bschoenmaeckers commented:
yes I will crash at:

`Caused by: java.lang.NullPointerException
    at net.minecraft.entity.Entity.isInvisibleToPlayer(Entity.java:2420)
    at net.minecraft.client.renderer.entity.RenderLivingBase.canRenderName(RenderLivingBase.java:506)
    at net.minecraft.client.renderer.entity.RenderLiving.canRenderName(RenderLiving.java:26)
    at net.minecraft.client.renderer.entity.RenderLiving.canRenderName(RenderLiving.java:16)
    at net.minecraft.client.renderer.entity.RenderLivingBase.renderName(RenderLivingBase.java:488)
    at net.minecraft.client.renderer.entity.RenderLivingBase.renderName(RenderLivingBase.java:25)
    at net.minecraft.client.renderer.entity.Render.doRender(Render.java:66)
    at net.minecraft.client.renderer.entity.RenderLivingBase.doRender(RenderLivingBase.java:212)
    at net.minecraft.client.renderer.entity.RenderLiving.doRender(RenderLiving.java:51)
    at net.minecraft.client.renderer.entity.RenderLiving.doRender(RenderLiving.java:16)
    at net.minecraft.client.renderer.entity.RenderManager.doRenderEntity(RenderManager.java:371)`

and more...

I know it isn't up to you to fix this and I know my code is  hacky as hell and that it makes unexpected situations for other mods. There are not a lot of options for me, I can call  the world load event manually but that will make the code a lot more messier and will probably cause a lot more incompatibilities.

elifoster commented:
I would recommend taking a look at how bspkrsCore does it.

bschoenmaeckers commented:
the original code comes from bspkrsCore but bspkrs pulled it out of bspkrsCore because of the compatibility issues. I just updated his code and added some more features to it.

covers1624 commented:
i have to dead to school but i would rather kick your issue in the head rather than avert it in 1 mod if you know what i mean.

bschoenmaeckers commented:
Do you have a better solution for our problem?

covers1624 commented:
I have some ideas, but still working on it, ill let you know.

bschoenmaeckers commented:
Ok tnx

covers1624 commented:
Ok, 2 options.
1, ASM isInvisibleToPlayer to return false if player is null.
2, In your render, set the player to your fake player, render your stuff, set to null again.

bschoenmaeckers commented:
I will go with option 2 because there will be more NPE and I dont want to ASM all of them.

covers1624 commented:
Sweet, Let me know how you get on.
